src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -0,0 +1,271 @@
+                                AbstractMemorySegmentImpl dst, long dstFromOffset, long dstToOffset) {
+        final long srcBytes = srcToOffset - srcFromOffset;
+        final long dstBytes = dstToOffset - dstFromOffset;
+        src.checkAccess(srcFromOffset, srcBytes, true);
+        dst.checkAccess(dstFromOffset, dstBytes, true);
+
+        final long bytes = Math.min(srcBytes, dstBytes);
+        final boolean srcAndDstBytesDiffer = srcBytes != dstBytes;
+
+        if (bytes == 0) {
+            return srcAndDstBytesDiffer ? 0 : -1;
+        } else if (bytes < NATIVE_THRESHOLD_MISMATCH) {
+            return mismatch(src, srcFromOffset, dst, dstFromOffset, 0, (int) bytes, srcAndDstBytesDiffer);
+        } else {
+            long i;
+            if (SCOPED_MEMORY_ACCESS.getByte(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset) !=
+                    SCOPED_MEMORY_ACCESS.getByte(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset)) {
+                return 0;
+            }
+            i = AbstractMemorySegmentImpl.vectorizedMismatchLargeForBytes(src.sessionImpl(), dst.sessionImpl(),
+                    src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset,
+                    dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset,
+                    bytes);
+            if (i >= 0) {
+                return i;
+            }
+            final long remaining = ~i;
+            assert remaining < 8 : "remaining greater than 7: " + remaining;
+            i = bytes - remaining;
+            return mismatch(src, srcFromOffset + i, dst, dstFromOffset + i, i, (int) remaining, srcAndDstBytesDiffer);
+        }
+    }
+
+    @ForceInline
+    private static long mismatch(AbstractMemorySegmentImpl src, long srcFromOffset,
+                                 AbstractMemorySegmentImpl dst, long dstFromOffset,
+                                 long start, int bytes, boolean srcAndDstBytesDiffer) {
+        final int limit = (int) (bytes & (NATIVE_THRESHOLD_MISMATCH - 8));
+        int offset = 0;
+        for (; offset < limit; offset += 8) {
+            final long s = SCOPED_MEMORY_ACCESS.getLong(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final long d = SCOPED_MEMORY_ACCESS.getLong(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
+                return start + offset + mismatch(s, d);
+            }
+        }
+        int remaining = bytes - offset;
+        // 0...0X00
+        if (remaining >= 4) {
+            final int s = SCOPED_MEMORY_ACCESS.getInt(src.sessionImpl(), src.unsafeGetBase(), src.unsafeGetOffset() + srcFromOffset + offset);
+            final int d = SCOPED_MEMORY_ACCESS.getInt(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + dstFromOffset + offset);
+            if (s != d) {
```

Comment:
Can we run `mismatch(s, d)` and then check the returned index (e.g. `!= 0`) instead of doing a full comparison and then another mismatch? Or is that slower?

Reply:
Assuming the most likely outcome is the values `s` and `d` are the same, I think the current way is faster. We avoid xor and counting bits which we likely do not have any use for.